forms-flow-documents/src/formsflow_documents/resources/pdf.py

```diff
@@ -0,0 +1,140 @@
+import string
+from http import HTTPStatus
+
+from flask import current_app, make_response, render_template, request
+from flask_restx import Namespace, Resource
+from formsflow_api_utils.exceptions import BusinessException
+from formsflow_api_utils.utils import (
+    CLIENT_GROUP,
+    REVIEWER_GROUP,
+    auth,
+    cors_preflight,
+    profiletime,
+)
+from werkzeug.utils import secure_filename
+
+from formsflow_documents.services import PDFService
+
+API = Namespace("Form", description="Form")
+
+
+@API.route("/<string:form_id>/submission/<string:submission_id>/render", doc=False)
+class FormResourceRenderPdf(Resource):
+    """Resource to render form and submission details as html."""
+
+    @staticmethod
+    @auth.require
+    @profiletime
+    def get(form_id: string, submission_id: string):
+        """Form rendering method."""
+        pdf_service = PDFService(form_id=form_id, submission_id=submission_id)
+        default_template = "index.html"
+        template_name = request.args.get("template_name")
+        template_variable_name = request.args.get("template_variable")
+        use_template = bool(template_name)
+
+        template_name = (
+            pdf_service.url_decode(secure_filename(template_name))
+            if use_template
+            else default_template
+        )
+
+        template_variable_name = (
+            pdf_service.url_decode(secure_filename(template_variable_name))
+            if template_variable_name
+            else None
+        )
+        if not pdf_service.search_template(template_name):
+            raise BusinessException(
+                "Template not found!", HTTPStatus.INTERNAL_SERVER_ERROR
+            )
+        if template_variable_name and not pdf_service.search_template(
+            template_variable_name
+        ):
+            raise BusinessException(
+                "Template variables not found!", HTTPStatus.INTERNAL_SERVER_ERROR
+            )
+
+        render_data = pdf_service.get_render_data(
+            use_template, template_variable_name, request.headers.get("Authorization")
+        )
+        headers = {"Content-Type": "text/html"}
+        return make_response(
+            render_template(template_name, **render_data), 200, headers
+        )
+
+
+@cors_preflight("POST,OPTIONS")
+@API.route(
+    "/<string:form_id>/submission/<string:submission_id>/export/pdf",
+    methods=["POST", "OPTIONS"],
+)
+@API.doc(
+    params={
+        "timezone": {
+            "description": "Timezone of client device eg: Asia/Calcutta",
+            "in": "query",
+            "type": "string",
+        }
+    }
+)
+class FormResourceExportPdf(Resource):
+    """Resource to export form and submission details as pdf."""
+
+    @staticmethod
+    @auth.require
+    @profiletime
+    def post(form_id: string, submission_id: string):
+        """PDF generation and rendering method."""
+        try:
+            if auth.has_one_of_roles([REVIEWER_GROUP, CLIENT_GROUP]):
```

Comment:
The library we use provides a decorator (https://github.com/thorwolpert/flask-jwt-oidc/blob/main/flask_jwt_oidc/jwt_manager.py#L213) to allow access to function only when the user has_one_of_roles. We could use it here, if there are no else condition to handle.

Reply:
👍